### PR TITLE
Implementation of setFluidProperties and GravityForcing in vc_ins_utilities class

### PR DESCRIPTION
**Pull request by Raamkrishnan**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format (i.e., `make indent`) run? For more information see
      `scripts/formatting/README.md`.
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

The new ```vc_ins_utilities``` class implements setting density and viscosity properties, as well as gravity force for the momentum equation. This eliminates the need for the setFluidProperties and GravityForcing classes in multiphase examples. Please take a look at ```vc_ins_utilities``` class and provide your suggestions on the design. 

Additionally, implementations from the LevelSetUtilities class have been used to replace the setLSProperties and TagLSRefinementCells classes in multiphase examples.

### CODE REVIEW

**Comment by amneetb** [comment]
@drwells : This PR removes 17,437 lines of code ;-)

File: include/ibamr/vc_ins_utilities.h
Comment (amneetb) [inline include/ibamr/vc_ins_utilities.h:44]: It should be `VCINSUtilities` instead of `VcNSUtilities`

File: src/utilities/vc_ins_utilities.cpp
Comment (amneetb) [inline src/utilities/vc_ins_utilities.cpp:30]: same here `VcINSUtilities`

File: src/utilities/vc_ins_utilities.cpp
Comment (amneetb) [inline src/utilities/vc_ins_utilities.cpp:45]: Maybe create an enum for TWO_PHASE and THREE_PHASE instead of string. You can place the enum in the header of this class.

File: src/utilities/vc_ins_utilities.cpp
Comment (amneetb) [inline src/utilities/vc_ins_utilities.cpp:772]: Also an enum for  FULL vs FLOW.

**Comment by amneetb** [comment]
@abarret : Can you take a look and suggest changes to @Raamkrishnan? Once this PR is merged he can remove repeated code files for the phase change examples and tests. 

@Raamkrishnan : The CI is failing. Can you take a look?

**Comment by abarret** [comment]
There's a lot of changes here, but they are good changes. Some initial suggestions:

1. Don't use an `enum` for the number of phases. `SetFluidProperties` already knows the number of phases based on the constructor, so you can make the number of phases an internal of the class. That way the class can have a uniform interface.
```c++
class SetFluidProperties
{
public:
SetFluidProperties(const std::string& object_name,
                   SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> adv_diff_solver,
                   SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_var,
                   const double rho_liquid,
                   const double rho_gas,
                   const double mu_liquid,
                   const double mu_gas,
                   const int ls_reinit_interval,
                   const double num_interface_cells);

SetFluidProperties(const std::string& object_name,
                   SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> adv_diff_solver,
                   SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_gas_var,
                   SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_solid_var,
                   const double rho_liquid,
                   const double rho_gas,
                   const double rho_solid,
                   const double mu_fluid,
                   const double mu_gas,
                   const double mu_solid,
                   const int ls_reinit_interval,
                   const double num_gas_interface_cells,
                   const double num_solid_interface_cells,
                   const bool set_mu_solid);

void setDensityPatchData()
{
    if (num_phases == 2)
        setDensityPatchData2PhaseFlows();
    else if (num_phases == 3)
        setDensityPatchData3PhaseFlows();
}
void setViscosityPatchData()
{
    if (num_phases == 2)
        setViscosityPatchData2PhaseFlows();
    else if (num_phases == 3)
        setViscosityPatchData3PhaseFlows();
}

private:
void setDensityPatchData2PhaseFlows();
void setDensityPatchData3PhaseFlows();
void setViscosityPatchData2PhaseFlows();
void setViscosityPatchData3PhaseFlows();

int d_num_phases;
}
```
2. Remove the raw pointers from the examples. You can get the same effect with regular objects.
```c++
IBAMR::LevelSetUtilities::SetLSProperties setSetLSProperties("SetLSProperties", level_set_gas_ops);
adv_diff_integrator->registerResetFunction(phi_var_gas,
                                           &IBAMR::LevelSetUtilities::setLSDataPatchHierarchy,
                                           static_cast<void*>(&setSetLSProperties));
```

**Comment by Raamkrishnan** [comment]
> There's a lot of changes here, but they are good changes. Some initial suggestions:
> 
> 1. Don't use an `enum` for the number of phases. `SetFluidProperties` already knows the number of phases based on the constructor, so you can make the number of phases an internal of the class. That way the class can have a uniform interface.
> 
> ```c++
> class SetFluidProperties
> {
> public:
> SetFluidProperties(const std::string& object_name,
>                    SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> adv_diff_solver,
>                    SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_var,
>                    const double rho_liquid,
>                    const double rho_gas,
>                    const double mu_liquid,
>                    const double mu_gas,
>                    const int ls_reinit_interval,
>                    const double num_interface_cells);
> 
> SetFluidProperties(const std::string& object_name,
>                    SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> adv_diff_solver,
>                    SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_gas_var,
>                    SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_solid_var,
>                    const double rho_liquid,
>                    const double rho_gas,
>                    const double rho_solid,
>                    const double mu_fluid,
>                    const double mu_gas,
>                    const double mu_solid,
>                    const int ls_reinit_interval,
>                    const double num_gas_interface_cells,
>                    const double num_solid_interface_cells,
>                    const bool set_mu_solid);
> 
> void setDensityPatchData()
> {
>     if (num_phases == 2)
>         setDensityPatchData2PhaseFlows();
>     else if (num_phases == 3)
>         setDensityPatchData3PhaseFlows();
> }
> void setViscosityPatchData()
> {
>     if (num_phases == 2)
>         setViscosityPatchData2PhaseFlows();
>     else if (num_phases == 3)
>         setViscosityPatchData3PhaseFlows();
> }
> 
> private:
> void setDensityPatchData2PhaseFlows();
> void setDensityPatchData3PhaseFlows();
> void setViscosityPatchData2PhaseFlows();
> void setViscosityPatchData3PhaseFlows();
> 
> int d_num_phases;
> }
> ```
> 
>  2. Remove the raw pointers from the examples. You can get the same effect with regular objects.
> 
> ```c++
> IBAMR::LevelSetUtilities::SetLSProperties setSetLSProperties("SetLSProperties", level_set_gas_ops);
> adv_diff_integrator->registerResetFunction(phi_var_gas,
>                                            &IBAMR::LevelSetUtilities::setLSDataPatchHierarchy,
>                                            static_cast<void*>(&setSetLSProperties));
> ```

Thank you @abarret. I have incorporated these suggestions. I also replaced pointer with objects in all the multiphase examples. Please take a look.

File: src/utilities/vc_ins_utilities.cpp
Comment (abarret) [inline src/utilities/vc_ins_utilities.cpp:651]: This looks suspicious to me. Why is one set with `dx` and one with `dy`?

File: src/utilities/vc_ins_utilities.cpp
Comment (abarret) [inline src/utilities/vc_ins_utilities.cpp:793]:
```suggestion
        using InterpolationTransactionComponent = HierarchyGhostCellInterpolation::InterpolationTransactionComponent;
```

**Review by abarret — CHANGES_REQUESTED** [review]
Comments on the library code. I'll look at the examples next.

File: include/ibamr/vc_ins_utilities.h
Comment (Raamkrishnan) [reply include/ibamr/vc_ins_utilities.h]: I removed ```ls_reinit_interval``` from SetFluidProperties class.

File: include/ibamr/vc_ins_utilities.h
Comment (Raamkrishnan) [reply include/ibamr/vc_ins_utilities.h]: I have two different constructors now. Since the ```grav_type``` option is unique for each constructor, I removed ```enum``` for ```grav_type``` and set it in the constructor directly.

File: src/utilities/vc_ins_utilities.cpp
Comment (Raamkrishnan) [reply src/utilities/vc_ins_utilities.cpp]: Thank you. I have now modified the code to use the correct implementation.

File: src/utilities/vc_ins_utilities.cpp
Comment (Raamkrishnan) [reply src/utilities/vc_ins_utilities.cpp]: I corrected it.

File: src/utilities/vc_ins_utilities.cpp
Comment (Raamkrishnan) [reply src/utilities/vc_ins_utilities.cpp]: From the literature, we explored different methods for setting densities at the side center. We want to have all of the options in one place. The default option works well for many multiphase problems. We can keep it as it is for now.

I have moved the flags closer to where they are checked.

**Comment by Raamkrishnan** [comment]
@abarret I have made changes in the code. Please take a look. Thank you for your suggestions.

File: examples/vc_navier_stokes/ex3/manufactured_soln_sec_6.1_simple_fcns.nb
Comment (abarret) [inline examples/vc_navier_stokes/ex3/manufactured_soln_sec_6.1_simple_fcns.nb:1]: Suggest saving as a `.wl`, `.m`, or `.wls` file to get rid of the meta data.

**Review by abarret — CHANGES_REQUESTED** [review]
This is my last batch of suggestions. Make sure you fill out the rest of the checklist. In particular, add a changelog.

File: src/utilities/vc_ins_utilities.cpp
Comment (Raamkrishnan) [reply src/utilities/vc_ins_utilities.cpp]: I have removed the ```grav_type``` parameter from both the code and input file. Now, users can choose between two different options for ```grav_type``` by calling different constructors.

File: include/ibamr/vc_ins_utilities.h
Comment (abarret) [inline include/ibamr/vc_ins_utilities.h:287]: `grav_type` is no longer a parameter to the system.
```suggestion
/*!
 * \brief The GravityForcing class provides an implementation of gravity force.
 * This class can be utilized to set the gravity force \f$ \rho g \f$ throughout the entire domain or specifically within the liquid and gas regions (excluding the solid phase)
 */
```

File: include/ibamr/vc_ins_utilities.h
Comment (abarret) [inline include/ibamr/vc_ins_utilities.h:299]:
```suggestion
    /*!
     * \brief Constructor for this class. Applies the gravitational force throughout the computational domain with the density retrieved from the provided <code>ins_hierarchy_integrator<\code>.
     *
     * grav_const stores the acceleration due to gravity.
     *
     */
    GravityForcing(const std::string& object_name,
                   SAMRAI::tbox::Pointer<INSVCStaggeredHierarchyIntegrator> ins_hierarchy_integrator,
                   std::vector<double> grav_const);
```

File: include/ibamr/vc_ins_utilities.h
Comment (abarret) [inline include/ibamr/vc_ins_utilities.h:312]: Make sure this is correct. Where is the gravity force specifically applied? How are `rho_neg` and `rho_pos` used?
```suggestion
    /*!
     * \brief Constructor for this class. Applies the gravitational force to only the portion of domain where the level set is negative.
     *
     * @param input_db provides parameters such as rho_neg, rho_pos, and num_gas_interface_cells.
     * grav_const stores the acceleration due to gravity.
     *
     */
    GravityForcing(const std::string& object_name,
                   SAMRAI::tbox::Pointer<AdvDiffHierarchyIntegrator> adv_diff_hierarchy_integrator,
                   SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_gas_var,
                   SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db,
                   std::vector<double> grav_const);
```

**Review by abarret — CHANGES_REQUESTED** [review]
Need to fix the `GravityForcing` comments and I'll approve.

@drwells do you want to look at this?

File: include/ibamr/vc_ins_utilities.h
Comment (Raamkrishnan) [reply include/ibamr/vc_ins_utilities.h]: I have verified this. The gravity force is applied across the entire domain. However, the density used in the gravity term only accounts for the densities of the liquid and gas phases, excluding the solid density. Specifically, rho_pos and rho_neg represent the densities of the liquid and gas phases, respectively.

**Comment by Raamkrishnan** [comment]
@abarret We can merge this branch if all looks good.

File: include/ibamr/vc_ins_utilities.h
Comment (abarret) [inline include/ibamr/vc_ins_utilities.h:84]: Comment should go above class definition.
```suggestion
/*!
 * \brief Class SetFluidProperties is a utility class which sets the fluid and
 * solid Eulerian density based on the current level set information
 */
class SetFluidProperties
{
```

**Review by abarret — APPROVED** [review]
Not sure how this was missed, but fix this one last thing. Looks good to me.

**Comment by Raamkrishnan** [comment]
Thank you @abarret for your suggestions.

**Comment by abarret** [comment]
@drwells do you want to look at this? Otherwise, I will merge.

**Review by drwells — APPROVED** [review]
It is used in tests - close enough. Nice work!

**Pull request merged**
